recover.go

```diff
@@ -0,0 +1,109 @@
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"log"
+	"net/http"
+	"runtime"
+)
+
+// A recoverMux recovers from panics during an http.Handler, sending an error
+// response to a client in the event a panic occurs.
+type recoverMux struct {
+	mux *http.ServeMux
+
+	DumpStack bool // set to true to dump the stack to the client on a panic
+}
+
+// NewRecoverMux returns a new recoverMux, creating and wrapping a new
+// http.ServeMux.
+func newRecoverMux() *recoverMux {
+	return &recoverMux{
+		mux: http.NewServeMux(),
+	}
+}
+
+// HandleFunc wraps the caller's handler with a recovery handler which recovers
+// from panics in the caller's handler. Response data is only written if the
+// caller's handler completes without panicking.
+func (rmux *recoverMux) HandleFunc(pattern string, handler func(http.ResponseWriter, *http.Request)) {
+	f := func(w http.ResponseWriter, r *http.Request) {
+		defer func() {
+			if x := recover(); x != nil {
+				rmux.recoverHandler(w, r)
+			}
+		}()
+		rw := newResponseWriter(w)
+		handler(rw, r)
+		err := rw.complete()
+		if err != nil {
+			log.Printf("error completing handler (URL %s): %s", r.URL, err.Error())
+		}
+	}
+	rmux.mux.HandleFunc(pattern, f)
+}
+
+// ServeHTTP uses the wrapped http.ServeMux to serve recoverable handlers.
+func (rmux *recoverMux) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	rmux.mux.ServeHTTP(w, r)
+}
+
+// RecoverHandler is the handler invoked when the client's handler panics.
+func (rmux *recoverMux) recoverHandler(w http.ResponseWriter, r *http.Request) {
+	fmt.Fprintf(w, "Something went wrong.\n")
+
+	if rmux.DumpStack {
+		var s [4096]byte
+		n := runtime.Stack(s[:], false)
```

Comment:
should use `debug.Stack` instead, both for ease, and ensuring you get the whole stack

Reply:
After looking at https://golang.org/pkg/net/http/#ServeMux, it looks like the standard library wants to initialize this itself and simply give you a pointer to a struct with unexported fields. It doesn't look like initializing and embedding it will work without copying it which seems bad.